content/post/2019-06-18.md

```diff
@@ -0,0 +1,256 @@
+
+Import data reading the CSV file and run it row by row to store the nodes and connections in the hash tables.
+
+Related entities can appear multiple times, as a source or as a target. When an entity appears for the first time, it is stored in the hash_nodes table. Attributes are associated to source entities and null attributes to target entities. It is a criterion that assumes this algorithm, but there could be others. If an entity appears first as a target as source, the null attributes are replaced by theirs as source.
+
+For each entity, the number of total links (hash_links), the number of inbound links (hash_links_in) and the number of outbound links (hash_links) are counted. This is done to allow ordering the nodes from greater to lesser degree when generating the file in GDF format.
+
+For each origin-target entity pair, the number of times that relation appears (hash_connections) and the attributes (hash_connections_attrib) are stored. In the first case, we get the weight of the relationship and the second we get the associated attributes.
+
+```r
+table_csv <-read_csv2(name_file_csv)
+num_rows=nrow(table_csv)
+num_cols=ncol(table_csv)
+for (i in 1:num_rows)
+{ 
+  node_source <- table_csv[[i,source]]
+  node_target <- table_csv[[i,target]]
+  print (i)
+  node_source_attribs<-null_attrib
+  for (j in 1:num_attribs)
+  { 
+   k=source_attribs[[j]]
+   raw_attrib<-table_csv[[i,k]]
+   cooked_attrib<-str_replace_all(raw_attrib,",","-") 
+   node_source_attribs[[j]]<-cooked_attrib
+  }
+  if (!(has.key(node_source, hash_nodes)))
+  {
+   hash_nodes[[node_source]] <- node_source_attribs
+   hash_links[[node_source]] <-0
+   hash_links_in[[node_source]] <-0
+   hash_links_out[[node_source]] <-0
+  }
+  else
+  {
+   node_source_attribs_old=hash_nodes[[node_source]]
+   if (identical(node_source_attribs_old,null_attrib)) 
+     {hash_nodes[[node_source]] <- node_source_attribs}
+  }
+  if (!(is.na(node_target)))
+  {
+   if (!(has.key(node_target, hash_nodes)))
+   {
+    hash_nodes[[node_target]] <- null_attrib
+    hash_links[[node_target]] <-0
+    hash_links_in[[node_target]] <-0
+    hash_links_out[[node_target]] <-0
+   }
+   par_nodes=paste(node_source,node_target)
+   if (!(has.key(par_nodes, hash_connections)))
+   {
+    hash_connections[[par_nodes]] <- 0
+    hash_connections_attrib[[par_nodes]]<-node_source_attribs
+   }
+   hash_connections[[par_nodes]] <- hash_connections[[par_nodes]] +1
+   hash_links[[node_source]] <- hash_links[[node_source]]+1
+   hash_links_out[[node_source]] <- hash_links_out[[node_source]]+1
+   hash_links[[node_target]] <-hash_links[[node_target]]+1
+   hash_links_in[[node_target]] <-hash_links_in[[node_target]]+1
+  }
+}
+```
+
+## Sort data according to connections
+
+The hash table object does not have the sort method, but has one to convert a hash table into a list. Once we have converted the hash list_links and list_connections into a list, we sort them down by number of connections.
```

Comment:
Yo pondría, para evitar la repetición, un "IT" en la primera frase:

... but has one to convert IT into a list.

Reply:
ok